README
Simply python swapsyncer 
 

debian/changelog
swapsyncer (0.1-ubuntu12) trusty; urgency=medium

  * add new site for sync = http://hacktech.xyz/swap/

 -- Zen aka Chibiko <zenogears.aka@gmail.com>  Mon, 06 Feb 2017 23:29:18 +0300

swapsyncer (0.1-ubuntu11) trusty; urgency=medium

  * fix bugs

 -- Zen aka Chibiko <zenogears.aka@gmail.com>  Tue, 15 Nov 2016 20:08:47 +0300

swapsyncer (0.1-ubuntu10) trusty; urgency=medium

  * change pyzenity to pyQt5; some bugs is here!

 -- SBT-Asonov-OV <sbt-asonov-ov@lin-361521aov.sigma.sbrf.ru>  Tue, 15 Nov 2016 13:40:11 +0300

swapsyncer (0.1-ubuntu9) trusty; urgency=medium

  * add progressbar O_o

 -- Zen aka Chibiko <zenogears.aka@gmail.com>  Sun, 13 Nov 2016 15:30:56 +0300

swapsyncer (0.1-ubuntu8) trusty; urgency=medium

  * fix some issues

 -- Zen aka Chibiko <zenogears.aka@gmail.com>  Mon, 22 Aug 2016 09:37:37 +0300

swapsyncer (0.1-ubuntu7) trusty; urgency=medium

  * fix some bugs + new functional + tests

 -- Zen aka Chibiko <zenogears.aka@gmail.com>  Sun, 21 Aug 2016 21:46:53 +0300

swapsyncer (0.1-ubuntu6) trusty; urgency=medium

  * bugfixes

 -- Zen aka Chibiko <zenogears.aka@gmail.com>  Tue, 13 Oct 2015 20:22:27 +0300

swapsyncer (0.1-trusty4) trusty; urgency=medium

  * bugfixes

 -- Zen aka Chibiko <zenogears.aka@gmail.com>  Tue, 13 Oct 2015 20:22:27 +0300

swapsyncer (0.1-trusty3) trusty; urgency=medium

  * iconfix

 -- Zen aka Chibiko <zenogears.aka@gmail.com>  Sun, 26 Jul 2015 03:58:05 +0300

swapsyncer (0.1-trusty2) trusty; urgency=medium

  * some updates with pythonzenity

 -- Zen aka Chibiko <zenogears.aka@gmail.com>  Sun, 26 Jul 2015 03:15:58 +0300

swapsyncer (0.1-trusty1) trusty; urgency=medium

  * Initial release. (Closes: #000001)

 -- Zen aka Chibiko <zenogears.aka@gmail.com>  Thu, 18 Dec 2014 22:15:36 +0300


usr/share/swapsyncer/test1.py
#!/usr/bin/python2
# -*- coding: utf-8 -*-

import sys
import urllib2

from PyQt4 import QtCore, QtGui

class DownloadThread(QtCore.QThread):

    data_downloaded = QtCore.pyqtSignal(object)

    def __init__(self, url):
        QtCore.QThread.__init__(self)
        self.url = url

    def run(self):
        info = urllib2.urlopen(self.url).info()
        self.data_downloaded.emit('%s\n%s' % (self.url, info))


class MainWindow(QtGui.QWidget):
    def __init__(self):
        super(MainWindow, self).__init__()
        self.list_widget = QtGui.QListWidget()
        self.button = QtGui.QPushButton("Start")
        self.button.clicked.connect(self.start_download)
        layout = QtGui.QVBoxLayout()
        layout.addWidget(self.button)
        layout.addWidget(self.list_widget)
        self.setLayout(layout)

    def start_download(self):
        urls = ['http://google.com', 'http://twitter.com', 'http://yandex.ru',
                'http://stackoverflow.com/', 'http://www.youtube.com/']
        self.threads = []
        for url in urls:
            downloader = DownloadThread(url)
            downloader.data_downloaded.connect(self.on_data_ready)
            self.threads.append(downloader)
            downloader.start()

    def on_data_ready(self, data):
        print (data)
        self.list_widget.addItem(unicode(data))


if __name__ == '__main__':
    app = QtGui.QApplication(sys.argv)
    window = MainWindow()
    window.resize(640, 480)
    window.show()
    sys.exit(app.exec_())

usr/share/swapsyncer/swapsyncer.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import os
import requests
import json
import sys
import pythonzenity
import getpass
import time

from datetime import datetime
from decimal import Decimal

from pythonzenity import Question
from pythonzenity import Warning
from pythonzenity import GetDirectory
from pythonzenity import InfoMessage
from pythonzenity import TextInfo
from pythonzenity import Progress

## Debug
debug_ON = 1

def debug(anything):
    if debug_ON:
        print(anything)

##Site
url = "http://hacktech.xyz/swap/"

conf_dir_path = "/home/" + getpass.getuser() + "/.swapsyncer"
debug("Config directory path: {0}".format(conf_dir_path))

conf_path = conf_dir_path + "/config.ini"
log_path = conf_dir_path + "/swapsyncer.log"

debug("Config path: {0}\nLog Path: {1}".format(conf_path,log_path))

dir_path = os.path.abspath(os.path.dirname(sys.argv[0]))

now = datetime.now()
date = datetime.strftime(datetime.now(), "%Y.%m.%d %H:%M:%S")
debug("Date: {0}".format(date))
out_string = ""

## Create conf dir if not exist
if not os.path.exists(conf_dir_path):
    debug("Create config directory")
    os.makedirs(conf_dir_path)

## create log and conf files if not exist  
lp = open(log_path, 'a+')
fp = open(conf_path, 'a+')

## create conf. file
if os.stat(conf_path).st_size == 0:
    if (Question(text="Конфиг не найден.\nДиректория SWAP не определена.\nХотите указать директорию под SWAP?")):
        swapdir = GetDirectory(sep=None)
        fp.write(str(swapdir)[2:-2])
        fp.close()

#if not lp.read():
#  lp.write(str(date))
#  lp.close()

fp = open(conf_path)
lp = open(log_path, 'w+')

eboots_path = str(fp.read())
fp.close()
   
if not os.path.exists(eboots_path):
    debug("Something gone wrong")
    Warning("Директория SWAP не найдена")

else:
    local_eboots = ([ f for f in os.listdir(eboots_path) if f.endswith('.bin') and os.path.isfile(os.path.join(eboots_path,f)) ])
    r = requests.get(url  + '/list.php')
    remote_eboots = r.json()
    download_list = list(set(remote_eboots) - set(local_eboots))
    upload_list = list(set(local_eboots) - set(remote_eboots))

    if len(download_list) == 0:
            out_string += "Нечего скачивать.\n"
    else:
        lp.write(str(date) + "\n")
        InfoMessage(text=str(len(download_list)) + " файлов будет скачано")
        testparam = float(100/float(len(download_list)))
        update = Progress(text='Downloading...', percentage=0, width=470, auto_close=True)
    i = 1;
    palka = 0;

    for fn in download_list:
        lp.write("D[ {0} / {1} ] {2}".format(i,len(download_list),fn))
        debug("Download: [ {0} / {1} ] {2}".format(i,len(download_list),fn))
        i+=1
        palka = round(float(palka) + float(testparam),2)
        files_to_download = len(download_list) - i
        datchik = int(palka)
        update(datchik,'Download: {0} % \tFiles: {1}'.format(palka,files_to_download))
        r = requests.get(url + "/SWAP/" + fn)
        with open(eboots_path + '/' + fn, 'wb') as fd:
            fd.write(r.content)
            fd.close()

    if len(upload_list) == 0:
        out_string += "Нечего закачивать.\n"
    else:
        InfoMessage(text=str(len(upload_list)) + " файлов будет загружено")
        testparam = float(100/float(len(upload_list)))
        update = Progress(text='Uploading...', percentage=0, width=470, auto_close=True)
        lp.write(str(date) + "\n")

    i = 1;
    palka = 0;

    for fn in upload_list:
        lp.write ("U[" + str(i) + "/" + str(len(upload_list)) + "]" + fn + "\n")
        i+=1
        palka = float(palka) + float(testparam)
        datchik = int(palka)
        update(datchik,"Загружаем данные на сервер, подождите... Это может занять длительное время. Загружено: " + str("%.1f" % palka) + " %")
        r = requests.post(url + "/uploadify.php", files={'Filedata': open(eboots_path + '/' + fn, 'rb')})
    lp.close();

    if len(out_string) > 0:
        InfoMessage(text=out_string)
    if len(upload_list)>0 or len(download_list)>0:
      if (Question(text="Желаете посмотреть изменения?")):
        TextInfo(filename=log_path, editable=False, width=400, height=200)


usr/share/swapsyncer/pyQT5_test1.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import sys
import os
import requests
import json
import sys
import pythonzenity
import getpass
import time

from datetime import datetime
from decimal import Decimal
from PyQt5.QtWidgets import (QMainWindow, QTextEdit, QWidget,
    QAction, QFileDialog, QApplication, QProgressBar, QPushButton, qApp, QHBoxLayout, QFrame, QSplitter, QLabel)
from PyQt5.QtGui import QIcon
from PyQt5.QtCore import QBasicTimer
from PyQt5.QtCore import Qt

debug_ON = True

def debug(anything):
    if debug_ON:
        print(anything)

##Site
url = "http://swapbase.ru/"

conf_dir_path = "/home/" + getpass.getuser() + "/.swapsyncer"
debug("Config directory path: {0}".format(conf_dir_path))

conf_path = conf_dir_path + "/config.ini"
log_path = conf_dir_path + "/swapsyncer.log"

debug("Config path: {0}\nLog Path: {1}".format(conf_path,log_path))

dir_path = os.path.abspath(os.path.dirname(sys.argv[0]))

now = datetime.now()
date = datetime.strftime(datetime.now(), "%Y.%m.%d %H:%M:%S")
debug("Date: {0}".format(date))
out_string = ""
mycind = True

## Create conf dir if not exist
if not os.path.exists(conf_dir_path):
    debug("Create config directory")
    os.makedirs(conf_dir_path)

## create log and conf files if not exist  
lp = open(log_path, 'a+')
fp = open(conf_path, 'a+')

## create conf. file
if os.stat(conf_path).st_size == 0:
    mycind = False

#if not lp.read():
#  lp.write(str(date))
#  lp.close()

fp = open(conf_path)
lp = open(log_path, 'w+')

eboots_path = str(fp.read())
fp.close()

class Example(QMainWindow):

    def __init__(self):
        super().__init__()
        self.initUI()


    def initUI(self):

        if not os.path.exists(eboots_path):
            debug("Something gone wrong")
            self.showDialog()

        else:
            self.local_eboots = ([ f for f in os.listdir(eboots_path) if f.endswith('.bin') and os.path.isfile(os.path.join(eboots_path,f)) ])
            r = requests.get(url  + '/list.php')
            self.remote_eboots = r.json()
            self.download_list = list(set(self.remote_eboots) - set(self.local_eboots))
            self.upload_list = list(set(self.local_eboots) - set(self.remote_eboots))
            #debug("{0}".format(len(self.download_list),self.download_list))

        openDir = QAction(QIcon('/mnt/Video/debianpack/swapsyncer/usr/share/icons/hicolor/24x24/apps/openx24.png'), 'Open SWAP directory', self)
        openDir.setShortcut('Ctrl+O')
        openDir.setStatusTip('Open SWAP directory')
        openDir.triggered.connect(self.showDialog)

        exitAction = QAction(QIcon('/mnt/Video/debianpack/swapsyncer/usr/share/icons/hicolor/24x24/apps/exitx24.png'), 'Exit', self)
        exitAction.setShortcut('Ctrl+Q')
        exitAction.setStatusTip('Exit')
        exitAction.triggered.connect(qApp.quit)

        self.toolbar = self.addToolBar('Open')
        self.toolbar.addAction(openDir)

        self.toolbar = self.addToolBar('Exit')
        self.toolbar.addAction(exitAction)

        self.pbar = QProgressBar(self)
        self.pbar.setGeometry(30, 40, 300, 25)

        if len(self.download_list) > 0:
            self.btn1 = QPushButton('Download: {0:6} files'.format(len(self.download_list)), self)
        else:
            self.btn1 = QPushButton('Nothing to download', self)
            self.pbar.setValue(100)

        self.btn1.move(100, 80)
        self.btn1.resize(150, 30)
        self.btn1.clicked.connect(self.doActionDownload)

        self.step1 = 0

        self.ubar = QProgressBar(self)
        self.ubar.setGeometry(30, 120, 300, 25)

        if len(self.upload_list) > 0:
            self.btn2 = QPushButton('Upload: {0:6} files'.format(len(self.upload_list)), self)
        else:
            self.btn2 = QPushButton('Nothing to upload', self)
            self.ubar.setValue(100)

        self.step2 = 0

        
        self.btn2.move(100, 160)
        self.btn2.resize(150, 30)
        self.btn2.clicked.connect(self.doActionUpload)
        

        self.lbl1 = QLabel('SWAP directory:', self)
        self.lbl1.move(10, 190)
        self.lbl1.resize(400, 15)
        if eboots_path:
            self.lbl2 = QLabel(eboots_path, self)
            self.lbl2.resize(400, 15)
            #debug(eboots_path)
            self.lbl2.move(10, 210)
        else:
            self.lbl1 = QLabel('Please, add SWAP directory', self)
            self.lbl1.move(10, 190)
            self.lbl1.resize(400, 15)

        self.setGeometry(300, 300, 350, 240)
        self.setWindowTitle('ToolBar')
        self.show()


    def showDialog(self):

        self.eboots_path = QFileDialog.getExistingDirectory(self, 'Open directory', '/home')
        if self.eboots_path:
            with open(conf_path, 'w') as fp:
                fp.write(str(self.eboots_path))
                #debug(self.eboots_path)
            sys.exit()

    def doActionDownload(self):
        if self.download_list is not None and len(self.download_list) > 0:
            self.i = 1;
            self.palka = 0;

            for fn in self.download_list:
                lp.write("D[ {0} / {1} ] {2}".format(self.i,len(self.download_list),fn))
                debug("Download: [ {0} / {1} ] {2}".format(self.i,len(self.download_list),fn))
                self.i+=1
                self.palka = round(float(self.palka) + float(100/float(len(self.download_list))),4)
                self.files_to_download = len(self.download_list) + 1 - self.i
                self.btn1.setText('Downloading: {0:6}'.format(self.files_to_download))

                if self.files_to_download > 0:
                    self.pbar.setValue(self.palka)
                    #debug(self.palka)
                else:
                    self.pbar.setValue(100)
                    self.btn1.setText('Nothing to download')

                r = requests.get(url + "/SWAP/" + fn)
                with open(eboots_path + '/' + fn, 'wb') as fd:
                    fd.write(r.content)
                    fd.close()

            self.download_list = None

        else:
            self.pbar.setValue(100)
            self.btn1.setText('Nothing to download')


    def doActionUpload(self):

        if self.upload_list is not None and len(self.upload_list) > 0:
            self.i = 1;
            self.palka = 0;

            for fn in self.upload_list:
                lp.write("U[ {0} / {1} ] {2}".format(self.i,len(self.upload_list),fn))
                debug("Upload: [ {0} / {1} ] {2}".format(self.i,len(self.upload_list),fn))
                self.i+=1
                self.palka = round(float(self.palka) + float(100/float(len(self.upload_list))),4)
                self.files_to_upload = len(self.upload_list) + 1 - self.i
                self.btn2.setText('Uploading: {0:6}'.format(self.files_to_upload))

                if self.files_to_upload > 0:
                    self.ubar.setValue(self.palka)
                    #debug(self.palka)
                else:
                    self.ubar.setValue(100)
                    self.btn2.setText('Nothing to upload')

                r = requests.post(url + "/uploadify.php", files={'Filedata': open(eboots_path + '/' + fn, 'rb')})

            self.upload_list = None

        else:
            self.ubar.setValue(100)
            self.btn2.setText('Nothing to upload')

if __name__ == '__main__':

    app = QApplication(sys.argv)
    ex = Example()
    sys.exit(app.exec_())

usr/share/swapsyncer/swapsyncer-qt.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import sys
import os
import requests
import json
import sys
import pythonzenity
import getpass
import time

from datetime import datetime
from decimal import Decimal
from PyQt5.QtWidgets import (QMainWindow, QTextEdit, QWidget,
    QAction, QFileDialog, QApplication, QProgressBar, QPushButton, qApp, QHBoxLayout, QFrame, QSplitter, QLabel)
from PyQt5.QtGui import QIcon
from PyQt5.QtCore import QBasicTimer
from PyQt5.QtCore import Qt

debug_ON = True

def debug(anything):
    if debug_ON:
        print(anything)

##Site
url = "http://hacktech.xyz/swap/"

conf_dir_path = "/home/" + getpass.getuser() + "/.swapsyncer"
debug("Config directory path: {0}".format(conf_dir_path))

conf_path = conf_dir_path + "/config.ini"
log_path = conf_dir_path + "/swapsyncer.log"

debug("Config path: {0}\nLog Path: {1}".format(conf_path,log_path))

dir_path = os.path.abspath(os.path.dirname(sys.argv[0]))

now = datetime.now()
date = datetime.strftime(datetime.now(), "%Y.%m.%d %H:%M:%S")
debug("Date: {0}".format(date))
out_string = ""
mycind = True

## Create conf dir if not exist
if not os.path.exists(conf_dir_path):
    debug("Create config directory")
    os.makedirs(conf_dir_path)

## create log and conf files if not exist  
lp = open(log_path, 'a+')
fp = open(conf_path, 'a+')

## create conf. file
if os.stat(conf_path).st_size == 0:
    mycind = False

#if not lp.read():
#  lp.write(str(date))
#  lp.close()

fp = open(conf_path)
lp = open(log_path, 'w+')

eboots_path = str(fp.read())
fp.close()

class Example(QMainWindow):

    def __init__(self):
        super().__init__()
        self.initUI()


    def initUI(self):

        if not os.path.exists(eboots_path):
            debug("Something gone wrong")
            self.showDialog()

        else:
            self.local_eboots = ([ f for f in os.listdir(eboots_path) if f.endswith('.bin') and os.path.isfile(os.path.join(eboots_path,f)) ])
            r = requests.get(url  + '/list.php')
            self.remote_eboots = r.json()
            self.download_list = list(set(self.remote_eboots) - set(self.local_eboots))
            self.upload_list = list(set(self.local_eboots) - set(self.remote_eboots))
            #debug("{0}".format(len(self.download_list),self.download_list))

        openDir = QAction(QIcon('/usr/share/icons/hicolor/24x24/apps/openx24.png'), 'Open SWAP directory', self)
        openDir.setShortcut('Ctrl+O')
        openDir.setStatusTip('Open SWAP directory')
        openDir.triggered.connect(self.showDialog)

        exitAction = QAction(QIcon('/usr/share/icons/hicolor/24x24/apps/exitx24.png'), 'Exit', self)
        exitAction.setShortcut('Ctrl+Q')
        exitAction.setStatusTip('Exit')
        exitAction.triggered.connect(qApp.quit)

        self.toolbar = self.addToolBar('Open')
        self.toolbar.addAction(openDir)

        self.toolbar = self.addToolBar('Exit')
        self.toolbar.addAction(exitAction)

        self.pbar = QProgressBar(self)
        self.pbar.setGeometry(30, 40, 300, 25)

        if len(self.download_list) > 0:
            self.btn1 = QPushButton('Download: {0:6} files'.format(len(self.download_list)), self)
        else:
            self.btn1 = QPushButton('Nothing to download', self)
            self.pbar.setValue(100)

        self.btn1.move(100, 80)
        self.btn1.resize(150, 30)
        self.btn1.clicked.connect(self.doActionDownload)

        self.step1 = 0

        self.ubar = QProgressBar(self)
        self.ubar.setGeometry(30, 120, 300, 25)

        if len(self.upload_list) > 0:
            self.btn2 = QPushButton('Upload: {0:6} files'.format(len(self.upload_list)), self)
        else:
            self.btn2 = QPushButton('Nothing to upload', self)
            self.ubar.setValue(100)

        self.step2 = 0

        
        self.btn2.move(100, 160)
        self.btn2.resize(150, 30)
        self.btn2.clicked.connect(self.doActionUpload)
        

        self.lbl1 = QLabel('SWAP directory:', self)
        self.lbl1.move(10, 190)
        self.lbl1.resize(400, 15)
        if eboots_path:
            self.lbl2 = QLabel(eboots_path, self)
            self.lbl2.resize(400, 15)
            #debug(eboots_path)
            self.lbl2.move(10, 210)
        else:
            self.lbl1 = QLabel('Please, add SWAP directory', self)
            self.lbl1.move(10, 190)
            self.lbl1.resize(400, 15)

        self.setGeometry(300, 300, 350, 240)
        self.setWindowTitle('SwapSyncer')
        self.show()


    def showDialog(self):

        self.eboots_path = QFileDialog.getExistingDirectory(self, 'Open SWAP directory', '/home')
        if self.eboots_path:
            with open(conf_path, 'w') as fp:
                fp.write(str(self.eboots_path))
                #debug(self.eboots_path)
            sys.exit(app.exec_())

    def doActionDownload(self):
        if self.download_list is not None and len(self.download_list) > 0:
            self.i = 1;
            self.palka = 0;

            for fn in self.download_list:
                lp.write("D[ {0} / {1} ] {2}".format(self.i,len(self.download_list),fn))
                debug("Download: [ {0} / {1} ] {2}".format(self.i,len(self.download_list),fn))
                self.i+=1
                self.palka = round(float(self.palka) + float(100/float(len(self.download_list))),4)
                self.files_to_download = len(self.download_list) + 1 - self.i
                self.btn1.setText('Downloading: {0:6}'.format(self.files_to_download))

                if self.files_to_download > 0:
                    self.pbar.setValue(self.palka)
                    #debug(self.palka)
                else:
                    self.pbar.setValue(100)
                    self.btn1.setText('Nothing to download')

                r = requests.get(url + "/SWAP/" + fn)
                with open(eboots_path + '/' + fn, 'wb') as fd:
                    fd.write(r.content)
                    fd.close()

            self.download_list = None

        else:
            self.pbar.setValue(100)
            self.btn1.setText('Nothing to download')


    def doActionUpload(self):

        if self.upload_list is not None and len(self.upload_list) > 0:
            self.i = 1;
            self.palka = 0;

            for fn in self.upload_list:
                lp.write("U[ {0} / {1} ] {2}".format(self.i,len(self.upload_list),fn))
                debug("Upload: [ {0} / {1} ] {2}".format(self.i,len(self.upload_list),fn))
                self.i+=1
                self.palka = round(float(self.palka) + float(100/float(len(self.upload_list))),4)
                self.files_to_upload = len(self.upload_list) + 1 - self.i
                self.btn2.setText('Uploading: {0:6}'.format(self.files_to_upload))

                if self.files_to_upload > 0:
                    self.ubar.setValue(self.palka)
                    #debug(self.palka)
                else:
                    self.ubar.setValue(100)
                    self.btn2.setText('Nothing to upload')

                r = requests.post(url + "/uploadify.php", files={'Filedata': open(eboots_path + '/' + fn, 'rb')})

            self.upload_list = None

        else:
            self.ubar.setValue(100)
            self.btn2.setText('Nothing to upload')

if __name__ == '__main__':

    app = QApplication(sys.argv)
    ex = Example()
    sys.exit(app.exec_())